### PR TITLE
support unix domain sockets as host in grpc_client

### PR DESCRIPTION
**Pull request by PleasantMachine9**
This allows communication with a grpc server that is listening
on the HTTP/2-over-UDS scheme.

UDS sockets compared with loopback TCP:
* UDS needs no loopback IP interface
* UDS can use filesystem-based access control
* UDS has no TCP/IP stack encapsulation overhead
  (higher throughput, less CPU use depending on use-case)
* For UDS, the HTTP/2 pseudo-header `:authority` needs a workaround,
  as the hostname is not a string but a tuple, and it may contain
  nul or slash characters which are not valid to send as values.
  The grpc-go client implementation sets the `:authority` to
  the hardcoded string `"localhost"`, so that behavior is followed,
  see https://github.com/grpc/grpc-go/pull/3730 and https://github.com/grpc/grpc-go/pull/4079

If the `{local, UnixSockPath}` tuple is passed along
from grpcbox to chatterbox, then the functionality
is working well, so there is not much impact on grpcbox aside
from updating the `:authority` field appropriately.

Note: the UDS usage is not documented in `chatterbox`,
their spec says that the host must be type `string()`
but it causes no issues with their library if that is not the case,
as they just pass the value through to `gen_tcp`
which supports the unix socket notation.

### CODE REVIEW

File: src/grpcbox.app.src [hunk 1]
```diff
@@ -9,7 +9,7 @@
                  acceptor_pool,
                  gproc,
                  ctx]},
-  {env, [{client, #{channels => [%% {default_channel, round_robin, [{http, "localhost", 8080, []}], #{}}
```
Comment (PleasantMachine9) [inline src/grpcbox.app.src:12]: Fixed this example, the round_robin option should have been inside the last argument (`#{}`) but it is the default value anyway so just removed it.

**Comment by PleasantMachine9** [comment]
# Testing the grpcbox client

I used the grpc/examples/python/helloworld:

*  https://github.com/grpc/grpc/blob/master/examples/protos/helloworld.proto coped into the `./proto` dir, then ran `rebar3 grpc gen proto/helloworld.proto`, 
* Used server: https://github.com/grpc/grpc/blob/master/examples/python/helloworld/greeter_server.py and changed
  `server.add_insecure_port('[::]:50051')` 
  to
  `server.add_insecure_port('unix-abstract:greeter_server')`
  before starting it
* set the default channel to be the unix abstract socket in `./src/grpcbox.app.src`
  ```
  {client, #{channels => [ {default_channel, [{http, {local, [0 | "greeter_server"]}, 0, []}], #{}}
  ]}}
  ```

*  then connected from erlang using `rebar 3 shell`
   ```
   helloworld_greeter_client:say_hello(#{name => "erlang"}).
   ```

Result:
```
1> helloworld_greeter_client:say_hello(#{name => "erlang"}).
{ok,#{message => <<"Hello, erlang!">>},
    #{headers =>
          #{<<":status">> => <<"200">>,
            <<"accept-encoding">> => <<"identity,gzip">>,
            <<"grpc-accept-encoding">> => <<"identity,deflate,gzip">>},
      trailers => #{}}}
```

**Comment by PleasantMachine9** [comment]
> src/grpcbox_subchannel.erl

Since the server side does not work with unix sockets (in this PR), not sure how this could be tested in-repo.

**Comment by PleasantMachine9** [comment]
On topic of testing difficulties in-repo, I did make an experimental commit for the server-side support too out of curiosity, but it seems quite a bit more complicated than client side because socket listener options are very different compared to TCP/IP. Notably
* anything referencing "ip" is not applicable and so has to be made optional (and handled when they are undefined)
* `gen_tcp` is told to listen on UDS by the special socket option `ifaddr` with the tuple `{ifaddr, {local, UnixPath}}`, as opposed to the current option `{ip, IpAddr}`

In any case, for viewing, that experiment is at https://github.com/PleasantMachine9/grpcbox/commit/817ccfed73db84c22be6fcf78b1ceff82acbdf4a but I think it would be a separate PR anyhow.

**Review by tsloughter — APPROVED** [review]

**Pull request merged**